File: exercises/quarkus/shipping-service/src/main/java/com/example/shipping/domain/identifier/ShippingContextKey.java
package com.example.shipping.domain.identifier;

import com.example.workshop.observability.DomainIdentifier;

import java.util.Objects;

public enum ShippingContextKey {

    ORDER_ID("order.id"),
    CUSTOMER_ID("customer.id"),
    SHIPMENT_ID("shipment.id");

    private final String key;

    ShippingContextKey(String key) {
        this.key = key;
    }

    public String key() {
        return key;
    }

    public DomainIdentifier of(String value) {
        Objects.requireNonNull(value, () -> name() + " value");
        return new SimpleIdentifier(key, value);
    }

    private record SimpleIdentifier(String key, String value)
            implements DomainIdentifier {
    }
}

File: exercises/dotnet/ShippingService/Domain/Identifiers/ShippingContextKey.cs
using SharedObservability;

namespace ShippingService.Domain.Identifiers;

/// <summary>
/// Shipping's typed identifier vocabulary. Shared wire names (order.id,
/// customer.id) are intentionally aligned with the Order service for
/// cross-service correlation.
/// </summary>
public sealed class ShippingContextKey
{
    public static readonly ShippingContextKey OrderId = new("order.id");
    public static readonly ShippingContextKey CustomerId = new("customer.id");
    public static readonly ShippingContextKey ShipmentId = new("shipment.id");

    public string Key { get; }

    private ShippingContextKey(string key) => Key = key;

    /// <summary>
    /// Bind this key to a concrete value, returning an <see cref="IDomainIdentifier"/>.
    /// </summary>
    public IDomainIdentifier Of(string value) => new BoundIdentifier(Key, value);

    private sealed record BoundIdentifier(string Key, string Value) : IDomainIdentifier;
}
